File: app/src/main/java/com/zack/enderweather/interactor/adapter/WeatherPagerAdapter.java
package com.zack.enderweather.interactor.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import com.zack.enderweather.model.bean.Weather;
import com.zack.enderweather.domain.fragment.WeatherFragment;

import java.util.List;

public class WeatherPagerAdapter extends FragmentStatePagerAdapter {

    private List<Weather> weatherList;

    public WeatherPagerAdapter(FragmentManager fm, List<Weather> weatherList) {
        super(fm);
        this.weatherList = weatherList;
    }

    @Override
    public Fragment getItem(int position) {
        return WeatherFragment.newInstance(weatherList.get(position));
    }

    @Override
    public int getCount() {
        return weatherList.size();
    }

    @Override
    public int getItemPosition(Object object) {
        return ((WeatherFragment) object).isDeleted() ? POSITION_NONE : super.getItemPosition(object);
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return weatherList.get(position).getBasicInfo().getCityName();
    }
}

File: app/src/main/java/net/zackzhang/code/haze/view/adapter/WeatherPagerAdapter.kt
package net.zackzhang.code.haze.view.adapter

import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import net.zackzhang.code.haze.model.DataManager
import net.zackzhang.code.haze.view.fragment.WeatherPageFragment

class WeatherPagerAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {

    override fun getItem(position: Int) = WeatherPageFragment.newInstance(position)

    override fun getCount() = DataManager.cityCount

    // 有城市的添加和删除，直接让所有已实例化的城市销毁重建，选择性更新会出问题
    // 实际上开销也不大，因为最多只会实例化三个城市
    override fun getItemPosition(obj: Any) = POSITION_NONE

    override fun getPageTitle(position: Int) = DataManager.getWeather(position).cityName
}
